params/include/alps/params/dict_types.hpp

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (C) 1998-2017 ALPS Collaboration. See COPYRIGHT.TXT
+ * All rights reserved. Use is subject to license terms. See LICENSE.TXT
+ * For use in publications, see ACKNOWLEDGE.TXT
+ */
+
+#ifndef ALPS_PARAMS_PARAM_TYPES_HPP_2b33f1b375e64b6fa9adcb68d7de2407
+#define ALPS_PARAMS_PARAM_TYPES_HPP_2b33f1b375e64b6fa9adcb68d7de2407
+
+#include <vector>
+#include <string>
+#include <stdexcept>
+#include <boost/mpl/vector/vector10.hpp>
+#include <boost/mpl/transform.hpp>
+#include <boost/mpl/front_inserter.hpp>
+#include <boost/mpl/copy.hpp>
+
+// forward declarations
+namespace alps{ namespace hdf5 {
+    class archive;
+}}
+
+namespace alps {
+    namespace params_ns {
+        namespace detail {
+            
+            // Have namespaces handy
+            namespace mpl=::boost::mpl;
+            namespace mplh=::boost::mpl::placeholders;
+
+            /// "Empty value" type
+            struct None {
```

Comment:
Yes, to represent of missing value.
Optional for every type would be distinguishing between missing int, missing string, missing double etc.
Is None really that annoying?

Reply:
>  Optional for every type would be distinguishing between missing int, missing string, missing double etc.

This would rather correspond to `boost::variant<boost::optional<int>, boost::optional<double>, ...>`.
Well, as I said, I don't fully realize all implications of using `boost::optional` instead of `None` (in either of the two forms).

Reply:
Sorry, was unwell and misread your comment. 
Yes, optional< variant<type1, type2,...> > might do, i will see how it fits. [Thanks! I now remember considering this for the previous version of params; for the life of me, can't recall why it did not work for me that time. It looks quite good now, and may simplify things.]
However, it's more a refactoring than a user-visible change, so may be of lower priority than fixing other issues.